### PR TITLE
Check cluster health during upgrade & disable shard allocation

### PR DESCRIPTION
**Pull request by carlcsaposs-canonical**

### CODE REVIEW

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 1]
```diff
@@ -913,6 +938,7 @@ def _stop_opensearch(self) -> None:
 
         if self.opensearch.is_node_up():
             # 1. Add current node to the voting + alloc exclusions
+            # TODO upgrade: disable on upgrade?
             self.opensearch_exclusions.add_current()
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:993]: > if stopping for upgrade, should we not add unit to exclusions?
We should

> also should we be doing this on normal restarts? e.g. restarting for tls or ip change
We also should because we're essentially avoiding opensearch to consider this node for new shards allocation

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 2]
```diff
@@ -953,6 +980,27 @@ def _upgrade_opensearch(self, event: _UpgradeOpenSearch) -> None:
...
+        # (Less dangerous if this unit fails to upgrade & reset the timeout—it's safer to have
+        # increased timeout than shard allocation disabled entirely, especially for long-running
+        # upgrades)
+        # Replaces "Disable shard allocation" step from
+        # https://www.elastic.co/guide/en/elastic-stack/8.13/upgrading-elasticsearch.html#rolling-upgrades
+        self.opensearch.request(
+            "PUT",
+            "/_all/_settings",
+            payload={"settings": {"index.unassigned.node_left.delayed_timeout": "5m"}},
+        )
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: You're right! What we can do here is: 
1. fetch the current setting value 
2. override it if it's < 5min, or keep as is otherwise
3. set it back to the original value if changed

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 3]
```diff
@@ -884,24 +897,42 @@ def _post_start_init(self, event: EventBase):
...
             self._put_monitoring_user()
 
         # clear waiting to start status
         self.status.clear(WaitingToStart)
 
+        if event.after_upgrade:
+            health = self.health.apply(wait_for_green_first=True, app=False)
+            self.health.apply_for_unit_during_upgrade(health)
+            # TODO upgrade: warning on step 9 https://www.elastic.co/guide/en/elastic-stack/8.13/upgrading-elasticsearch.html#rolling-upgrades
+            if health != HealthColors.GREEN:
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: I assume yes

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 4]
```diff
@@ -884,24 +897,86 @@ def _post_start_init(self, event: EventBase):
...
+                from https://www.elastic.co/guide/en/elastic-stack/8.13/upgrading-elasticsearch.html#upgrading-elasticsearch
+                """
+                if health_ == HealthColors.GREEN:
+                    return True
+                # Units that are on an upgraded version
+                # (It is not sufficient to check if this unit is first to upgrade. For example, in
+                # a 3 unit deployment, units 2 and 1 were upgraded. Then, the user does
+                # `juju refresh` to rollback. Units 0 and 2 are on version A and unit 1 is on
+                # version B. The cluster health could be yellow & we should proceed with rollback.)
+                upgraded_units = [
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:957]: It seems from the docs, that we should allow `YELLOW` - but NOT `YELLOW_TEMP`

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:993]: To summarize, upgrades go through the `stop` process which incorporates the allocation exclusion of the node.
The same stop process used in any operation requiring a restart, and which may not guarantee the same node comes back online due to an internal failure and whatnot. (TLS certificates rotation, IP changes etc...)
 
We do this for 2 reasons:
1. Prevent new shards fropm being allocated to the node
2. Trigger a shards relocation to other nodes (missing from our implementation the synchrnous wait for this to finish)

In short, we should exclude the currently upgrading node from the allocation. And since we're making use of `stop_opensearch` method, we're already doing it.

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: The idea I had was more with regards to the relation data. Not the `StoredState`.

---

Overriding it - without setting it back - is certainly inconvenient for the user who may have to reconfigure their indices settings, but definitely safer than disabling replica shards allocation in the case of a **long running** upgrade - which the Elastic docs doesn't really take into consideration.

I believe that was the conclusion we reached when we discussed the pros and cons of each. The major con of disabling replica shards allocation (as advised by the docs) is in the case of long upgrades - where no new replica shards **(of new indices)** will be created. (existing replicas will still be replicated)

---

I just made a research now, and I believe we must differentiate between a **cluster wide** `index.unassigned.node_left.delayed_timeout` and an index-level `delayed_timeout`.
The former only overrides the ones in the indices where no `delayed_timeout` has been set, **NOT** where it has been set.

---

Meaning that if we wanna go down this route, we should:
1. fetch the shards located in the node about to upgrade
2. retrieve the index names they belong to 
3. retrieve the delayed timeout for each - store it in the peer data - override it with `5m` (if 5m greater than whatever is set in there)
4. after upgrade reset it based on the relation data

---

Makes sense?

[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: sounds good

[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: it is the default behavior of the call you made above - it only affects the indices not explicitely configured by the user

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:957]: Addressed on MM

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 5]
```diff
@@ -884,24 +897,87 @@ def _post_start_init(self, event: EventBase):
...
+                    unit
+                    for unit in self._upgrade._sorted_units
+                    if self._upgrade._unit_workload_container_versions.get(unit.name)
+                    != self._upgrade._app_workload_container_version
+                ]
+                if len(upgraded_units) > 1:
+                    # More than 1 unit is upgraded
+                    return False
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:965]: I'm not entirely sure what we check here

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:993]: Addressed on MM

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 6]
```diff
@@ -955,6 +1033,29 @@ def _upgrade_opensearch(self, event: _UpgradeOpenSearch) -> None:
...
+        # upgrades)
+        # Replaces "Disable shard allocation" step from
+        # https://www.elastic.co/guide/en/elastic-stack/8.13/upgrading-elasticsearch.html#rolling-upgrades
+        self.opensearch.request(
+            "PUT",
+            "/_all/_settings",
+            payload={"settings": {"index.unassigned.node_left.delayed_timeout": "5m"}},
+        )
+
+        self.opensearch.request("POST", "/_flush", retries=3)
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:1056]:
```suggestion
         try:
             self.opensearch.request(
                 "POST", "/_flush?wait_if_ongoing=true", resp_status_code=True, retries=3
             )
         except OpenSearchHttpError:
             pass
```

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Carl! A few points left to address as discussed and should be good to go

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 7]
```diff
@@ -955,6 +1033,29 @@ def _upgrade_opensearch(self, event: _UpgradeOpenSearch) -> None:
...
+        # (Less dangerous if this unit fails to upgrade & reset the timeout—it's safer to have
+        # increased timeout than shard allocation disabled entirely, especially for long-running
+        # upgrades)
+        # Replaces "Disable shard allocation" step from
+        # https://www.elastic.co/guide/en/elastic-stack/8.13/upgrading-elasticsearch.html#rolling-upgrades
+        self.opensearch.request(
+            "PUT",
+            "/_all/_settings",
+            payload={"settings": {"index.unassigned.node_left.delayed_timeout": "5m"}},
+        )
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:1054]: This will fail, because of the inclusion of the .opendistro_security index in `_all` - which cannot be altered... 
We should be more cautious here, and change this to the following:
```suggestion
        indices = self.opensearch.request("GET", "/_settings?flat_settings=true")
        indices_to_update = ",".join([
            index
            for index, settings in indices.items()
            if not index.startswith(".") and "index.unassigned.node_left.delayed_timeout" not in settings
        ])
        if indices_to_update:
            self.opensearch.request(
                "PUT",
                f"/{indices_to_update}/_settings",
                payload={"settings": {"index.unassigned.node_left.delayed_timeout": 5m}},
            )
            
            # TODO: either store `indices_to_update` this in relation data 
            # OR: store them in an internal index.
```

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 8]
```diff
@@ -884,24 +897,87 @@ def _post_start_init(self, event: EventBase):
 
         self.node_lock.release()
 
-        self._upgrade.unit_state = upgrade.UnitState.HEALTHY
-        self._reconcile_upgrade()
+        # TODO upgrade: catch http errors
+        self.opensearch.request(
+            "PUT",
+            "/_all/_settings",
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:903]: We should fetch back the `indices_to_update` from wherever we store them

**Review by Mehdi-Bendriss — COMMENTED** [review]
I gave some thoughts to the `delayed_timeout` - we should be more selective

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: I gave it some more thoughts, we should be selective, please refer to [here](https://github.com/canonical/opensearch-operator/pull/261#discussion_r1583015342)

File: lib/charms/opensearch/v0/constants_charm.py [hunk 9]
```diff
@@ -42,6 +42,9 @@
 ClusterHealthYellow = (
     "1 or more 'replica' shards are not assigned, please scale your application up."
 )
+# TODO upgrade: add call to action for user?
+ClusterHealthRedUpgrade = "1 or more 'primary' shards are not assigned"
+ClusterHealthYellowUpgrade = "1 or more 'replica' shards are not assigned"
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/constants_charm.py:47]: maybe a "in the cluster" at the end?

Also, `red` means the cluster cannot recover a given index. You should add a "Cluster in RED" at the beginning

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 10]
```diff
@@ -884,24 +897,69 @@ def _post_start_init(self, event: EventBase):
...
+            # Reset to default value
+            payload={"settings": {"index.unassigned.node_left.delayed_timeout": None}},
```
Comment (phvalguima) [inline lib/charms/opensearch/v0/opensearch_base_charm.py]: I think we need to specify the level of `settings`. In the OS docs, it recommends a `persistent` setting:
```
PUT "/_cluster/settings?pretty"
{
    "persistent": {
        "cluster.routing.allocation.enable": "primaries"
    }
}
```

**Review by phvalguima — APPROVED** [review]

**Review by phvalguima — DISMISSED** [review]

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 11]
```diff
@@ -966,9 +1044,26 @@ def _upgrade_opensearch(self, event: _UpgradeOpenSearch) -> None:
...
+                "/_cluster/settings",
+                payload={"persistent": {"cluster.routing.allocation.enable": "primaries"}},
+            )
+        except OpenSearchHttpError:
+            logger.exception("Failed to disable shard allocation before upgrade")
+            self.node_lock.release()
+            event.defer()
+            return
+        try:
+            self.opensearch.request("POST", "/_flush", retries=3)
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:1060]:
```suggestion
            self.opensearch.request("POST", "/_flush?wait_if_ongoing=true", retries=3)
```

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thank you! I only have 1 comment on the flush request, but LGTM

**Review by Mehdi-Bendriss — APPROVED** [review]

**Review by phvalguima — APPROVED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]
